samplekotlin/src/main/java/com/example/chattutorial/ChannelActivity3.kt

```diff
@@ -19,69 +22,67 @@ import io.getstream.chat.android.livedata.controller.ChannelController
         val messageListViewModel: MessageListViewModel by viewModels { factory }
         val messageInputViewModel: MessageInputViewModel by viewModels { factory }
 
-        // set custom AttachmentViewHolderFactory
+        // Set custom AttachmentViewHolderFactory
         messageListView.setAttachmentViewHolderFactory(MyAttachmentViewHolderFactory())
 
-        // step 2 = we bind the view and ViewModels. they are loosely coupled so its easy to customize
+        // Step 2 - Bind the view and ViewModels, they are loosely coupled so it's easy to customize
         channelHeaderViewModel.bindView(channelHeaderView, this)
         messageListViewModel.bindView(messageListView, this)
         messageInputViewModel.bindView(messageInputView, this)
 
-        // step 3 - let the message input know when we open a thread
-        messageListViewModel.mode.observe(this) {
-            when (it) {
-                is MessageListViewModel.Mode.Thread -> messageInputViewModel.setActiveThread(it.parentMessage)
-                is MessageListViewModel.Mode.Normal -> messageInputViewModel.resetThread()
+        // Step 3 - Let the message input know when we open a thread
+        messageListViewModel.mode.observe(this) { mode ->
+            when (mode) {
+                is Thread -> messageInputViewModel.setActiveThread(mode.parentMessage)
+                is Normal -> messageInputViewModel.resetThread()
             }
         }
 
-        // step 4 - handle navigate up state
-        messageListViewModel.state.observe(this) {
-            when (it) {
-                is MessageListViewModel.State.NavigateUp -> finish()
+        // Step 4 - Handle navigate up state
+        messageListViewModel.state.observe(this) { state ->
+            if (state is NavigateUp) {
+                finish()
             }
         }
 
-        // step 5 - let the message input know when we are editing a message
+        // Step 5 - Let the message input know when we are editing a message
         messageListView.setOnMessageEditHandler {
             messageInputViewModel.editMessage.postValue(it)
         }
 
-        // step 6 - handle back button behaviour correctly when you're in a thread
+        // Step 6 - Handle back button behaviour correctly when you're in a thread
         channelHeaderView.onBackClick = {
             messageListViewModel.onEvent(MessageListViewModel.Event.BackButtonPressed)
         }
-
         onBackPressedDispatcher.addCallback(this) {
             channelHeaderView.onBackClick()
         }
 
-        // custom typing info header bar
+        // Custom typing info header bar
         val nobodyTyping = "nobody is typing"
-
         typingHeader.text = nobodyTyping
 
-        // obtain a channel controller
+        // Obtain a ChannelController
         val channelController: ChannelController =
             ChatDomain.instance().useCases.watchChannel(cid, 0).execute().data()
```

Comment:
Same in Kotlin sample - we should replace it with `getChannelController` use case